### PR TITLE
[DNM] Redesign InMemoryDataset data model with hyperrectangles selection

### PR DESCRIPTION
**Pull request by crusaderky**
**Note:** This PR is still missing some components - see checklist below.

`versioned_hdf5.wrappers.InMemoryDataset` is plagued by slowness caused by the control code, both because the control code itself is very slow and because it calls `h5py.Dataset.__getitem__` independently for every single chunk.

The new control logic introduced in this PR aims to minimise the number of calls to `h5py.Dataset.__getitem__` by selecting large rectangles of unmodified data spanning multiple chunks.

This new design introduces a helper class, `StagedChangesArray`, which handles all the control logic. This class has minimal external API and is completely abstracted from h5py.

There are also other major design improvements. Namely, in master `__setitem__` first loads all impacted chunks into memory, in the off chance that they are not wholly covered by the parameter index, and then updates them. In this PR, this happens exclusively for the chunks that are not wholly covered by the index of `__setitem__`.
In other words, `a[:] = 1` in master loads the whole dataset into memory if it's not in memory already, whereas in this PR it never touches the disk.

Analogously, in master `resize()` reads the whole dataset from disk. This no longer happens in this PR, which only loads the edge chunks if the shape is not exactly divisible by chunk_size.

## Status
- [x] implement `__getitem__`
- [x] implement `__setitem__`
- [x] implement `resize()`
- [x] crudely test `__getitem__` by hand
- [x] crudely test `__setitem__` by hand
- [x] crudely test `resize()` by hand
- [x] integrate with wrappers.py
- [x] all existing unit tests pass
- [ ] new unit tests to stress the new classes in isolation
- [ ] clean up legacy code
- [x] high level documentation
- [ ] break PR up into stages to simplify code review

## Demo
See `demo.ipynb`.

## Benchmarks
See `benchmarks.ipynb`.
Early benchmarks show up to 500x speedup in the control code and up to 2000x reduction in the number of calls to `h5py.Dataset.__getitem__`.

### CODE REVIEW

File: versioned_hdf5/subchunk_map.py [hunk 1]
```diff
@@ -45,113 +51,503 @@ def _smallest(x: Py_ssize_t, a: Py_ssize_t, m: Py_ssize_t) -> Py_ssize_t:
...
-    stop = _ceiling(stop - s_start, s_step) if stop > s_start else 0
+    We define a custom exception instead of IndexError as raising this is a normal
+    occourrence that will be caught quite a long way down the line and we don't want to
+    silence an accidental, genuine IndexError.
+    """
 
-    return slice(int(start), int(stop), 1)
 
+@cython.cclass
+class IndexChunkMapper:
```
Comment (crusaderky) [inline versioned_hdf5/subchunk_map.py:65]: This class and all of its subclasses are just a refactor of the legacy code, with no logic changes.
I've rewired the legacy `as_subchunk_map` to use the refactored code.
There are minor additions, which are not used by the legacy `as_subchunk_map`:
- new methods `chunks_indexer` and `whole_chunks_indexer`
- the method `chunks_submap` allows selecting more than one contiguous chunk at a time, but `as_subchunk_map` will only ever ask for one
- the same method raises an exception for early exit when none of the chunks requested is impacted by the index. This never happens with the legacy `as_subchunk_map` by definition.

File: versioned_hdf5/subchunk_map.py [hunk 2]
```diff
@@ -336,3 +674,861 @@ def as_subchunk_map(
         chunk_subidxs = chunk_subidxs[:idx_len]
 
         yield Tuple(*chunk_slices), arr_subidxs, chunk_subidxs
+
```
Comment (crusaderky) [inline versioned_hdf5/subchunk_map.py:677]: Refactor ends here. Everything below this line is new code that's not been integrated yet.
I must point out that the code cythonizes very poorly. This can be improved in a later iteration.

File: versioned_hdf5/subchunk_map.py [hunk 3]
```diff
@@ -336,3 +674,861 @@ def as_subchunk_map(
...
+        indices_i = np.broadcast_to(indices_i, states.shape)
+        # Get only the indices of modified chunks
+        indices_i = indices_i[mask]
+        indices.append(indices_i.tolist())
+
+    values_idx = (states[mask] - 1).tolist()
+    return [(tuple(idx), value_idx) for *idx, value_idx in zip(*indices, values_idx)]
+
+
+def _hypercubes_between(
```
Comment (crusaderky) [inline versioned_hdf5/subchunk_map.py:1384]: I'm not happy with this design. It performs particularly badly in some edge cases, e.g. it's great after you call
`a[::2 * chunk_size, :] = 1`
but it's awful after
`a[:, ::2 * chunk_size] = 1`
because in the latter case you'll end up loading one chunk at a time.

It's also problematically slow when you have most of the dataset staged in memory, e.g. after `a[:] = 1`

For the sake of moving things forward, I would like to deliver like this as is and rework it later on with a smarter algorithm that can generate long and narrow rectangles.

File: versioned_hdf5/subchunk_map.py [hunk 4]
```diff
@@ -336,3 +674,861 @@ def as_subchunk_map(
...
+                (idx_from[0] + 1,) + (0,) * (ndim - 1),
+                (idx_to[0], *shape[1:]),
+            )
+
+        for ni, nj in _hypercubes_between(
+            (-1,) * (ndim - 1),
+            idx_to[1:],
+            shape[1:],
+        ):
+            yield (idx_to[0], *ni), (idx_to[0] + 1, *nj)
```
Comment (crusaderky) [inline versioned_hdf5/subchunk_map.py:1438]: Yield, tuples, etc. are all things that compile in cython but give you ZERO performance benefit from pure python.
Hacks to make it stay in C space for longer are very much possible but at the cost of some complication.

File: versioned_hdf5/subchunk_map.py [hunk 5]
```diff
@@ -180,145 +576,87 @@ def as_subchunk_map(
...
-        chunk_subindexes.append(
-            [(chunk_slice, chunk_slice.raw, ()) for chunk_slice in chunk_slices_gen]
-        )
+        mapper = EverythingMapper(n, d)
+        mappers.append(mapper)
+
+    return idx, mappers
+
+
+def as_subchunk_map(
```
Comment (crusaderky) [inline versioned_hdf5/subchunk_map.py:612]: This is just a temporary refactor, for my sanity to verify that I didn't break anything fundamental early on, and for the sanity of reviewers. It will be completely removed in the final version.

**Comment by crusaderky** [comment]
Status update:

- integration done
- high level docs done
- most pre-existing tests are green (12 failures)

**Comment by crusaderky** [comment]
Status update:

- integration done
- high level docs done
- all pre-existing tests are green

**Comment by crusaderky** [comment]
First PR that breaks this down: https://github.com/deshaw/versioned-hdf5/pull/371

More will follow using a similar pattern. The idea is to reduce the size of this PR from 5500 lines to less than 500.

Timeline:
1. https://github.com/deshaw/versioned-hdf5/pull/371 (unused)
2. Refactor `as_subchunk_map` to be object-oriented; add more unit tests
3. (the big one) `StagedChangesArray` (unused; depends on 1 and 2). Includes high level docs.
4. Rewire wrappers to use StagedChangesArray instead of as_subchunk_map (depends on 3)

**Comment by ArvidJB** [comment]
I just tried this and am getting errors like
```
/codemill/bessen/fork/base/src/python/deshaw/shawcache/hdf5.py:8162: in _try_remove_unused_entry
    data_group[axis_name][:new_length] = data_group[axis_name][bool_index]
/codemill/bessen/versioned_hdf5_venv/lib64/python3.11/site-packages/versioned_hdf5/wrappers.py:1270: in __setitem__
    self.dataset.__setitem__(index, value)
h5py/_objects.pyx:54: in h5py._objects.with_phil.wrapper
    ???
h5py/_objects.pyx:55: in h5py._objects.with_phil.wrapper
    ???
/codemill/bessen/versioned_hdf5_venv/lib64/python3.11/site-packages/versioned_hdf5/wrappers.py:953: in __setitem__
    self.staged_changes[args] = val
subchunk_map.pyx:1158: in versioned_hdf5.subchunk_map.StagedChangesArray.__setitem__
    ???
/usr/local/python/python-3.11/std/lib64/python3.11/site-packages/numpy/lib/stride_tricks.py:413: in broadcast_to
    return _broadcast_to(array, shape, subok=subok, readonly=True)
/usr/local/python/python-3.11/std/lib64/python3.11/site-packages/numpy/lib/stride_tricks.py:345: in _broadcast_to
    if any(size < 0 for size in shape):
/usr/local/python/python-3.11/std/lib64/python3.11/site-packages/numpy/lib/stride_tricks.py:345: in <genexpr>
    if any(size < 0 for size in shape):
E   TypeError: '<' not supported between instances of 'NoneType' and 'int'
```

Also note that I got a build error
```
      ../versioned_hdf5/meson.build:58:3: ERROR: Include dir /usr/local/python/python-3.11/std/lib64/python3.11/site-packages/numpy/_core/include does not exist.
```
and had to change `_core` to `core`. This is with numpy version 1.24.4. Strangely this seems to work in the CI run?

**Comment by ArvidJB** [comment]
Here's a reproducer:
```
In [1]: with h5py.File('data.h5', 'w') as f:
   ...:     vf = VersionedHDF5File(f)
   ...:     with vf.stage_version('r0') as sv:
   ...:         sv.create_dataset('value', data=np.array([[[1, 2], [3, 4]], [[5, 6], [7, 8]]]), chunks=(10, 10, 10), maxshape=(None, None, None))
   ...:
[PYFLYBY] from versioned_hdf5 import VersionedHDF5File
[PYFLYBY] import h5py
[PYFLYBY] import numpy as np

In [2]: with h5py.File('data.h5', 'r+') as f:
   ...:     vf = VersionedHDF5File(f)
   ...:     with vf.stage_version('r1') as sv:
   ...:         sv['value'][:, 1:1, :] = np.zeros((2, 0, 2))
   ...:
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[2], line 4
      2 vf = VersionedHDF5File(f)
      3 with vf.stage_version('r1') as sv:
----> 4     sv['value'][:, 1:1, :] = np.zeros((2, 0, 2))

File /codemill/bessen/versioned_hdf5_venv/lib64/python3.11/site-packages/versioned_hdf5/wrappers.py:1270, in DatasetWrapper.__setitem__(self, index, value)
   1268     self.dataset = new_dataset
   1269     return
-> 1270 self.dataset.__setitem__(index, value)

File h5py/_objects.pyx:54, in h5py._objects.with_phil.wrapper()

File h5py/_objects.pyx:55, in h5py._objects.with_phil.wrapper()

File /codemill/bessen/versioned_hdf5_venv/lib64/python3.11/site-packages/versioned_hdf5/wrappers.py:953, in InMemoryDataset.__setitem__(self, args, val)
    950     mtype = None
    952 # === END CODE FROM h5py.Dataset.__setitem__ ===
--> 953 self.staged_changes[args] = val

File subchunk_map.pyx:1158, in versioned_hdf5.subchunk_map.StagedChangesArray.__setitem__()

File <__array_function__ internals>:200, in broadcast_to(*args, **kwargs)

File /usr/local/python/python-3.11/std/lib64/python3.11/site-packages/numpy/lib/stride_tricks.py:413, in broadcast_to(array, shape, subok)
    367 @array_function_dispatch(_broadcast_to_dispatcher, module='numpy')
    368 def broadcast_to(array, shape, subok=False):
    369     """Broadcast an array to a new shape.
    370
    371     Parameters
   (...)
    411            [1, 2, 3]])
    412     """
--> 413     return _broadcast_to(array, shape, subok=subok, readonly=True)

File /usr/local/python/python-3.11/std/lib64/python3.11/site-packages/numpy/lib/stride_tricks.py:345, in _broadcast_to(array, shape, subok, readonly)
    343 if not shape and array.shape:
    344     raise ValueError('cannot broadcast a non-scalar to a scalar array')
--> 345 if any(size < 0 for size in shape):
    346     raise ValueError('all elements of broadcast shape must be non-'
    347                      'negative')
    348 extras = []

File /usr/local/python/python-3.11/std/lib64/python3.11/site-packages/numpy/lib/stride_tricks.py:345, in <genexpr>(.0)
    343 if not shape and array.shape:
    344     raise ValueError('cannot broadcast a non-scalar to a scalar array')
--> 345 if any(size < 0 for size in shape):
    346     raise ValueError('all elements of broadcast shape must be non-'
    347                      'negative')
    348 extras = []

TypeError: '<' not supported between instances of 'NoneType' and 'int'
```

**Comment by ArvidJB** [comment]
This might be specific to arrays of size 0 (where one of the dimensions is of length 0)?

**Comment by ArvidJB** [comment]
There's a data corruption bug here:
```
In [1]: with h5py.File('data.h5', 'w') as f:
   ...:     vf = VersionedHDF5File(f)
   ...:     with vf.stage_version('r0') as sv:
   ...:         sv.create_dataset('value', data=np.array([[1, 2, 3, 4, 5, 6]]), chunks=(10, 1250), maxshape=(None, None))
   ...:
[PYFLYBY] from versioned_hdf5 import VersionedHDF5File
[PYFLYBY] import h5py
[PYFLYBY] import numpy as np

In [2]: for i in range(1, 31):
   ...:     with h5py.File('data.h5', 'r+') as f:
   ...:         vf = VersionedHDF5File(f)
   ...:         with vf.stage_version(f'r{i}') as sv:
   ...:             value_ds = sv['value']
   ...:             value_ds.resize((value_ds.shape[0] + 1, 6))
   ...:             value_ds[-1, :] = np.array([1, 2, 3, 4, 5, 6])
   ...:

In [3]: with h5py.File('data.h5', 'r') as f:
   ...:     vf = VersionedHDF5File(f)
   ...:     cv = vf[vf.current_version]
   ...:     print(cv['value'][:])
   ...:
[[0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [1 2 3 4 5 6]]
```

**Comment by ArvidJB** [comment]
The 11th update corrupts the data:
```
In [9]: with h5py.File('data.h5', 'w') as f:
   ...:     vf = VersionedHDF5File(f)
   ...:     with vf.stage_version('r0') as sv:
   ...:         sv.create_dataset('value', data=np.array([[1, 2, 3, 4, 5, 6]]), chunks=(10, 1250), maxshape=(None, None))
   ...:

In [10]: for i in range(1, 11):
    ...:     version_name = f'r{i}'
    ...:     print(version_name)
    ...:     with h5py.File('data.h5', 'r+') as f:
    ...:         vf = VersionedHDF5File(f)
    ...:         with vf.stage_version(version_name) as sv:
    ...:             value_ds = sv['value']
    ...:             value_ds.resize((value_ds.shape[0] + 1, 6))
    ...:             value_ds[-1, :] = np.array([1, 2, 3, 4, 5, 6])
    ...:     with h5py.File('data.h5', 'r') as f:
    ...:         vf = VersionedHDF5File(f)
    ...:         cv = vf[vf.current_version]
    ...:         print(cv['value'][:])
    ...:
r1
[[1 2 3 4 5 6]
 [1 2 3 4 5 6]]
r2
[[1 2 3 4 5 6]
 [1 2 3 4 5 6]
 [1 2 3 4 5 6]]
r3
[[1 2 3 4 5 6]
 [1 2 3 4 5 6]
 [1 2 3 4 5 6]
 [1 2 3 4 5 6]]
r4
[[1 2 3 4 5 6]
 [1 2 3 4 5 6]
 [1 2 3 4 5 6]
 [1 2 3 4 5 6]
 [1 2 3 4 5 6]]
r5
[[1 2 3 4 5 6]
 [1 2 3 4 5 6]
 [1 2 3 4 5 6]
 [1 2 3 4 5 6]
 [1 2 3 4 5 6]
 [1 2 3 4 5 6]]
r6
[[1 2 3 4 5 6]
 [1 2 3 4 5 6]
 [1 2 3 4 5 6]
 [1 2 3 4 5 6]
 [1 2 3 4 5 6]
 [1 2 3 4 5 6]
 [1 2 3 4 5 6]]
r7
[[1 2 3 4 5 6]
 [1 2 3 4 5 6]
 [1 2 3 4 5 6]
 [1 2 3 4 5 6]
 [1 2 3 4 5 6]
 [1 2 3 4 5 6]
 [1 2 3 4 5 6]
 [1 2 3 4 5 6]]
r8
[[1 2 3 4 5 6]
 [1 2 3 4 5 6]
 [1 2 3 4 5 6]
 [1 2 3 4 5 6]
 [1 2 3 4 5 6]
 [1 2 3 4 5 6]
 [1 2 3 4 5 6]
 [1 2 3 4 5 6]
 [1 2 3 4 5 6]]
r9
[[1 2 3 4 5 6]
 [1 2 3 4 5 6]
 [1 2 3 4 5 6]
 [1 2 3 4 5 6]
 [1 2 3 4 5 6]
 [1 2 3 4 5 6]
 [1 2 3 4 5 6]
 [1 2 3 4 5 6]
 [1 2 3 4 5 6]
 [1 2 3 4 5 6]]
r10
[[0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [1 2 3 4 5 6]]
```

**Comment by ArvidJB** [comment]
Looking at the design here it looks eerily similar to the ideas in https://github.com/deshaw/versioned-hdf5/pull/313...

Can we please implement this without any "plans" like `ChangesPlan` / `ResizePlan` / ...? We cannot track multiple upserts, deletes, resizes in a staged commit. That's a fundamental design change to versioned-hdf5 which I am not comfortable with.

**Comment by crusaderky** [comment]
Hi @ArvidJB ,
thanks for looking into this. Sorry about the bugs - they are to be expected, as I'm still in the middle of writing unit tests to properly shake them out. I apologise if I didn't communicate this correctly and as a result you were expecting a finished product.

>       ../versioned_hdf5/meson.build:58:3: ERROR: Include dir /usr/local/python/python-3.11/std/lib64/python3.11/site-packages/numpy/_core/include does not exist.

Fixed by removing `cimport numpy` altogether. I was already on the fence with it, as it gives a barely measurable benefit, so I just yanked it out.

> This might be specific to arrays of size 0 (where one of the dimensions is of length 0)?

Yes, this is definitely because it's an edge case that is untested in the preexisting unit tests.

> The 11th update corrupts the data:

It's the update that causes the shape to be exactly divisible by chunk_size, so it behaves slightly differently from the others.

These bugs all look straightforward and I'll have them fixed in a couple of hours.
Thorough hypothesis test coverage will follow in a matter of days, so that you (hopefully) won't find more afterwards.

> Can we please implement this without any "plans" like ChangesPlan / ResizePlan / ...? We cannot track multiple upserts, deletes, resizes in a staged commit. That's a fundamental design change to versioned-hdf5 which I am not comfortable with.

Hold on.
The plans _do not_ track multiple changes.
They are just a neat way to decouple decision-making logic from the data and make debugging easier, and nothing more.
What each individual `__setitem__` and `resize` method call does is
1. formulate a plan,
2. apply the plan and irreversibly change the state of the in-memory data,
3. discard the plan.

This is no different from mutating the `data_dict` like it happens in master.

I might have just as well designed this as follows:
1. `__getitem__`, `__setitem__`, `resize` tell a specialized function "this is my state and this is what the user wants; tell me what to do"
2. the function yields instructions, e.g. "take chunk at index 123 and resize it to size 10"
3. `resize` etc. implements the instruction, thus mutating the state, and then asks for the next instruction until no more are yielded

To clarify, this would be just a cosmetic change - within a single `__setitem__`, the decision-maker doesn't need to know the updated state between every instruction, because it just told the data-holding class what bits of the state to update. In other words you could generate each instruction and apply it or just dump the iterator into a list, generating all the instructions ahead of time, and then consume them all at once.

Would you prefer to have a face-to-face conversation about this so that we can be aligned on the way forward?

**Comment by crusaderky** [comment]
> Here's a reproducer:

Here's a great bit about the new design: you can very easily yank h5py out of your reproducers.
```python
>>> base = np.array([[[1, 2], [3, 4]], [[5, 6], [7, 8]]])
>>> s = StagedChangesArray(
...     base.__getitem__,
...     dtype=base.dtype,
...     shape=base.shape,
...     chunk_size=(10, 10, 10),
... )
>>> s[:, 1:1, :] = np.zeros((2, 0, 2))

TypeError: '<' not supported between instances of 'NoneType' and 'int'
```
Plans allow easily debugging:
```python
>>> s._setitem_plan((slice(None), slice(1, 1), slice(None)))
SetItemPlan<shape=None, 0 loads from base into 0 chunks, 0 appends of full chunks, 0 appends from __setitem__ value, 0 updates>
chunk_states:
[[[0]]]
```
shape was uninitialised due to early exit. One-liner fix: https://github.com/deshaw/versioned-hdf5/pull/370/commits/8728ed54602d164c5f1785226564041f586648e1

**Comment by crusaderky** [comment]
There are now 4 sub-PRs ready for review and merge:
- #371
- #372
- #374
- #373 (depends on 372 and 374)

**Comment by crusaderky** [comment]
This PR was abandoned as it was proven that virtual datasets are exceptionally slow in hdf5 C, so we had to redesign everything around it.
Superseded by #386.
